### ISSUE TITLE
Orchestrate the `replica-installer`

### ISSUE THREAD

**Issue by dmartinol** [RHEcosystemAppEng/SaaSi#15]
Do we have to model dependencies and ordering of installation in the `replica-installer`?

E.g., what if there are Jobs needed to run before installing the app? do we need to wait for the completion of this worker before installing the other resources? Aren't we re-implementing Helm chart lifecycle?

**Comment by zvigrinberg** [comment]
In my opinion, It's redundant to re-implement lifecycle hooks like in Helm, not only because we can re-use it with helm packages in Golang, but mainly because we need to be more generic, as still there are a lot of installations and deployments that weren't deployed using helm.

Three approaches to deal with that:

1. Don't try to impose an order of pods and jobs, if job needs to run before a pod, they will applied to cluster at the same time, and  then using an appropriate startupProbe And/Or LivenessProbe  that check for the Availability of the resource created/prepared by the job, eventually the pod will "catch up" with the job' result.
This way we can intervene by incrementing up initialDelay of the Startup/Liveness Probe , according to some logic, before applying manifests to the cluster.

2. Impose an order, we need to transform one-time jobs/pods' logic (that supposed to be run before application pods)  into init-containers of one of the application' pod, the work here(beside the transformation) is to decide how to recognize such order between resources in the source cluster:

- We can looks for helm hooks annotations.
- In addition, we can support a dedicated portion in the input yaml file, that will define some resources that should be executed before application pods, the resource will be specified by according to labels/annotations that the one time jobs/pods in the source cluster already has, and distinguish them from the application pods.

3. 2 phases process -   Maybe a better option from section 2 is to support a dedicated portion in the input yaml file, that will define some one-time resources that should be executed before application pods, like in previous section according to labels, but don't map it to init-containers, instead, get them and all their resources(configmap,secrets,RBAC, etc) etc, using labels, and apply them to the cluster before applying application resources(without the one-time resources), we can support a configurable parameter that will impose waiting between these 2 phases(in units of seconds)

**Comment by dmartinol** [comment]
Let's keep this point for a dedicated discussion. For now we can go ahead with a single, uncontrolled deployment of all the extracted resources, but in the next releases we have to think of a manageable way to control the deployment. WDYT?

**Comment by zvigrinberg** [comment]
Yeah make sense, we can stick to option 1 for the moment, this is the default anyway in the absence of a mechanism that will take care of it, so we'll defer the development of such mechanism to later stages.
